docs/reference/scripting/grok-syntax.asciidoc

```diff
@@ -0,0 +1,231 @@
+[[grok]]
+=== Grokking grok
+Grok is a regular expression dialect that supports reusable aliased expressions. Grok works really well with syslog logs, Apache and other webserver
+logs, mysql logs, and generally any log format that is written for humans and
+not computer consumption.
+
+Grok sits on top of the https://github.com/kkos/oniguruma/blob/master/doc/RE[Oniguruma] regular expression library, so any regular expressions are
+valid in grok. Grok uses this regular expression language to allow naming
+existing patterns and combining them into more complex patterns that match your
+fields.
+
+[[grok-syntax]]
+==== Grok patterns
+The {stack} ships with numerous https://github.com/elastic/elasticsearch/blob/master/libs/grok/src/main/resources/patterns/grok-patterns[predefined grok patterns] that simplify working with grok. The syntax for reusing grok patterns
+takes one of the following forms:
+
+[%autowidth]
+|===
+|`%{SYNTAX}` | `%{SYNTAX:SEMANTIC}` |`%{SYNTAX:SEMANTIC:TYPE}`
+|===
+
+`SYNTAX`::
+The name of the pattern that will match your text. For example, `NUMBER` and
+`IP` are both patterns that are provided within the default patterns set. The
+`NUMBER` pattern matches data like `3.44`, and the `IP` pattern matches data
+like `55.3.244.1`. 
+
+`SEMANTIC`::
+The identifier you give to the piece of text being matched. For  example, `3.44`
+could be the duration of an event, so you might call it `duration`. The string
+`55.3.244.1` might identify the `client` making a request.
+
+`TYPE`::
+The data type you want to cast your named field. `int`, `long`, `double`,
+`float` and `boolean` are supported types.
+
+For example, let's say you have message data that looks like this:
+
+[source,txt]
+----
+3.44 55.3.244.1
+----
+
+The first value is a number, followed by what appears to be an IP address. You
+can match this text by using the following grok expression:
+
+[source,txt]
+----
+%{NUMBER:duration} %{IP:client}
+----
+
+[[grok-patterns]]
+==== Use grok patterns in Painless scripts
+You can incorporate predefined grok patterns into Painless scripts to extract
+data. To test your script, use either the {painless}/painless-execute-api.html[field contexts] of the Painless execute API or create a runtime field that
+includes the script. Runtime fields offer greater flexibility and accept
+multiple documents, but the Painless execute API is a great option if you don't
+have write access on a cluster where you're testing a script.
+
+TIP: If you need help building grok patterns to match your data, use the 
+{kibana-ref}/xpack-grokdebugger.html[Grok Debugger] tool in {kib}.
+
+For example, if you're working with Apache log data, you can use the
+`%{COMMONAPACHELOG}` syntax, which understands the structure of Apache logs. A
+sample document might look like this:
+
+[source,txt]
+----
+"timestamp":"2020-04-30T14:30:17-05:00","message":"40.135.0.0 - - 
+[30/Apr/2020:14:30:17 -0500] \"GET /images/hm_bg.jpg HTTP/1.0\" 200 24736"
+----
```

Comment:
might be worth leaving a comment in the asciidoc for anyone that wonders why you have the line break. It certainly makes it more readable!

Reply:
++ will add a comment for future contributors 👍